### PR TITLE
[WIP] Outlier simulation and rejection v2

### PR DESCRIPTION
**Pull request by ArnavJain23**

### CODE REVIEW

File: drone_data_analyze.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from src.utils import DataAnalyzer
+from src.utils import Parser
+
+def main():
```
Comment (BauerJustin) [inline drone_data_analyze.py:4]: Just a note for both of us: let's figure out how to move this file and `plot_localization_history` into a separate `scripts` folder or something so there are only the two main files in the parent directory

File: src/simulator/linear_trajectory.py [hunk 2]
```diff
@@ -37,12 +37,17 @@ def __init__(self, anchors):
 
     def get_measurements(self):
         measurements = {}
+        realistic_outlier_injected = False
```
Comment (BauerJustin) [inline src/simulator/linear_trajectory.py:40]: I think we can remove this flag to keep it simple and allow the outlier to be injected possibly more than once in a single set of measurements. Can always lower the probability instead to lower the chances

File: src/utils/data_analyzer.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        except FileNotFoundError:
+            print("File not found. Please check the file path.")
+            return []
+        except json.JSONDecodeError as e:
+            raise Exception(f"Error decoding JSON: {e}")
+        
+        anchor_distances = []
+        for _, packet in data:
+            distances = packet["measurements"]
+            anchor_distances.append([distances.get("81", 0), distances.get("82", 0), distances.get("83", 0), distances.get("84", 0)])
```
Comment (BauerJustin) [inline src/utils/data_analyzer.py:38]: It's probably best to use the `Measurements` class throughout but not a requirement

**Review by BauerJustin — COMMENTED** [review]
My general comment is that we should simplify the outlier handling by creating a new class that drops individual measurements outside of the range. This should make the code less complicated and we can then disable the buffer and rely on the KF instead

File: src/algorithms/buffer.py [hunk 4]
```diff
@@ -1,13 +1,16 @@
...
 from src.utils import Position, Measurements
-
+import copy
 
 class Buffer:
-    def __init__(self, base_type, size, filter_outliers=False):
+    def __init__(self, base_type, size, filter_outliers=False, outlier_rejection_version="V2", outlier_replacement=False):
         self.base_type = base_type
         self.buffer = []
+        self.filtered_buffer = []
```
[the remark this answers is not in the document]
Comment (ArnavJain23) [reply src/algorithms/buffer.py:9]: Create new class for outlier rejection, remove filtered buffer, leave buffer code intact

File: src/algorithms/buffer.py [hunk 5]
```diff
@@ -16,38 +19,89 @@ def add(self, value):
...
+        new_val = np.array(self.buffer[-1].unpack())
+        if any(val < 0 or val > 5 for val in new_val):
+            if self.outlier_replacement:
+                replaced_measurements = self._replace_outlier(copy.copy(self.buffer[-1]))
+                if replaced_measurements is not None:
+                    self._add_to_filtered_buffer(replaced_measurements)
+        else:
+            self._add_to_filtered_buffer(self.buffer[-1])
+        
+    def _replace_outlier(self, measurements):
```
[the remark this answers is not in the document]
Comment (ArnavJain23) [reply src/algorithms/buffer.py:84]: Keep outlier replacement code, move it to outlier rejection class, keep experimenting and compare with KF replacement

File: src/constants.py [hunk 6]
```diff
@@ -1,7 +1,7 @@
 # Drone Tracker
-BASE = "pos"  # pos, measurement
+BASE = "measurement"  # pos, measurement
 
-FIRST_UPDATES_SKIPPED = 5
+FIRST_UPDATES_SKIPPED = 0
```
[the remark this answers is not in the document]
Comment (ArnavJain23) [reply src/constants.py:4]: Remove flag, algorithms and system are robust to enough to work with outliers in first few measurements

File: src/constants.py [hunk 7]
```diff
@@ -11,8 +11,11 @@
 BUFFER_SIZE = 20
 BUFFER_FILTER_OUTLIERS = True
 
+OUTLIER_REPLACEMENT_ENABLED = False
+OUTLIER_REJECTION_VERSION = "V2" # "V1", "V2"
```
[the remark this answers is not in the document]
Comment (ArnavJain23) [reply src/constants.py:15]: Remove OUTLIER_REJECTION_VERSION,  BUFFER_FILTER_OUTLIERS, disable buffer

File: src/constants.py [hunk 8]
```diff
@@ -11,8 +11,11 @@
...
+
 MEASURE_VARIANCE = True
-VARIANCE_SIZE = 5
+VARIANCE_SIZE = 20
```
[the remark this answers is not in the document]
Comment (ArnavJain23) [reply src/constants.py:18]: Set to 20 for testing, change back to 5, add measurement variance buffer

File: src/utils/data_analyzer.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        self.data_complete_without_outliers = self.remove_incomplete_and_outlier_data()
+        self.total_complete_data_without_outliers = len(self.data_complete_without_outliers)
+
+        self.extract_outlier_data_info()
+        self.extract_incomplete_data_info()
+        self.print_relevant_details()
+
+    def extract_outliers(self):
+        for distances in self.measurements:
+            if any(value < 0 or value > 5 for value in distances):
```
[the remark this answers is not in the document]
Comment (ArnavJain23) [reply src/utils/data_analyzer.py]: Add constants use anchor to determine max value

File: src/simulator/linear_trajectory.py (hunk 2 above)
Comment (ArnavJain23) [reply src/simulator/linear_trajectory.py:40]: Keep to ensure single outlier injection to largely resemble actual data

File: src/utils/data_analyzer.py (hunk 3 above)
Comment (ArnavJain23) [reply src/utils/data_analyzer.py]: Nice to have

**Pull request merged**